### ISSUE TITLE
findCirclesGrid() finding duplicate points

### ISSUE THREAD

**Issue by opencv-pushbot** [opencv/opencv#4775]
Transferred from http://code.opencv.org/issues/3774

```
|| Dennis Evangelista on 2014-06-27 15:59
|| Priority: Normal
|| Affected: 2.4.0 - 2.4.8
|| Category: calibration, 3d
|| Tracker: Bug
|| Difficulty: 
|| PR: 
|| Platform: Other / Mac OSX
```
## findCirclesGrid() finding duplicate points

```
Sometimes, 

retval,corners = findCirclesGrid(image,size,flag=cv2.CALIB_CV_CLUSTER) 

returns retval=True and corners but where the corners have duplicate points.  We can detect the duplicate points by checking for the pairwise distances using scipy.spatial.distance.pdist.  Removing the CALIB_CV_CLUSTER makes the duplicate points go away but results in fewer pattern detections. 

If this is expected behavior, it might help to note this in the documentation; inadvertently having duplicate points really messes up our calibration.
```
##### Dmitry Retinskiy on 2014-09-17 09:56

```
Ilya, could you check this?
Thanks.
-   Assignee set to Ilya Lysenkov
-   Status changed from New to Open
```

**Comment by MattClarkson** [comment]
Hi, is anyone available to look at this? Im using 2.4.11 and i believe its a valid issue in that software version.

**Comment by alalek** [comment]
Complete reproducer is required.

**Comment by MattClarkson** [comment]
![1479224227386188288_left](https://cloud.githubusercontent.com/assets/1152637/21285213/62fadd3e-c42a-11e6-803d-2649ecc3f184.png)

This gets a bit stranger. I was definitely getting duplicates, then I put some debug statements in, and recompiled a few times.

The image I was using is attached.

I was basically doing (pseudo code):
`
    cv::Mat greyImage;
    cv::cvtColor(image, greyImage, CV_BGR2GRAY);

      cv::Rect leftROI(0, 0, imageSize.width / 2, imageSize.height);
      cv::Mat leftImage = greyImage(leftROI);

      cv::Rect rightROI(imageSize.width / 2, 0, imageSize.width / 2, imageSize.height);
      cv::Mat rightImage = greyImage(rightROI);
`
and running findCircles on the left and right sub-image

`
  std::vector<cv::Point2f> circlesLeft;
  std::vector<cv::Point2f> circlesRight;
  unsigned int numberOfCircles = m_PatternSize.width * m_PatternSize.height;
  cv::SimpleBlobDetector::Params params;
  params.maxArea = 10e4;
  cv::Ptr<cv::FeatureDetector> blobDetector = new cv::SimpleBlobDetector(params);

    bool foundLeft = cv::findCirclesGrid(leftImage, m_PatternSize, circlesLeft, cv::CALIB_CB_SYMMETRIC_GRID | cv::CALIB_CB_CLUSTERING, blobDetector);

  bool foundRight = cv::findCirclesGrid(rightImage, m_PatternSize, circlesRight, cv::CALIB_CB_SYMMETRIC_GRID | cv::CALIB_CB_CLUSTERING, blobDetector);

`
and the left one worked reliably, and the right one produced duplicates.

But now I'm not 100% sure its reproducible. If someone else could try with the same image and code similar to the above, I'd be grateful.

**Comment by MattClarkson** [comment]
Comment updated. The rescaling seems to confuse the correct detection of circles in the right ROI. Thanks.

**Comment by MattClarkson** [comment]
OK, in more detail. The above image is 1920x540. Its obtained from a 1920x1080 interlaced camera, so we drop every alternate row to avoid comb artifacts. But we rescale it in the y direction to obtain the original number of rows, and so that camera calibration has the right y-scale factor.

I also tried, instead of using an ROI to select either the left or right sub-image, I simply cloned the greyImage twice into leftImage, rightImage and in leftImage I blanked/masked out the right image and in the rightImage I blanked/masked out the left image. 

However, the same problem remains. 

It was caused by incorrect re-ordering of the corners. In OpenCV 2.4.11, file circlesgrid.cpp, line 341 there is

`
  if(!isAsymmetricGrid)
  {
    double dist1 = norm(sortedCorners[0] - sortedCorners[1]);
    double dist2 = norm(sortedCorners[1] - sortedCorners[2]);

    if((dist1 > dist2 && patternSize.height > patternSize.width) || (dist1 < dist2 && patternSize.height < patternSize.width))
    {
      for(size_t i=0; i<sortedCorners.size()-1; i++)
      {
        sortedCorners[i] = sortedCorners[i+1];
      }
      sortedCorners[sortedCorners.size() - 1] = firstCorner;
    }
  }
`
This causes the corners to be incorrectly ordered, and then the following rectification fails, the mapping to the homography is wrong as a result, all leading to an incorrect flnn search, and duplicates. The above piece of code says if the width of the pattern is less than the height of the pattern, and patternSize.height  is less than patternSize.width (in my case, 7,10 respectively, so statement is true), then re-order the corners. This assumption doesn't take into account patterns at a large angle to the camera, and perspective foreshortening. If I remove this section, the duplicates disappear.
